polynote-frontend/polynote/ui/component/comment.ts

```diff
@@ -0,0 +1,517 @@
+"use strict";
+
+
+import {button, Content, div, span, tag, TagElement, textarea, textbox} from "../util/tags";
+import { CreateComment, CurrentIdentity, DeleteComment, UIMessageRequest, UIMessageTarget, UpdateComment } from "../util/ui_event";
+import * as uuidv4 from 'uuid/v4';
+import {CellComment} from "../../data/data";
+import {PosRange} from "../../data/result";
+import * as monaco from "monaco-editor";
+import {editor, IDisposable} from "monaco-editor";
+import IStandaloneCodeEditor = editor.IStandaloneCodeEditor;
+import IModelContentChangedEvent = editor.IModelContentChangedEvent;
+import TrackedRangeStickiness = editor.TrackedRangeStickiness;
+
+export type CommentID = string
+
+// to ensure UUID uniqueness, make sure to only create new comments with createCellComment!
+export function createCellComment({range, author, createdAt, content}: Omit<CellComment, 'uuid'>): CellComment {
+    const uuid = uuidv4();
+    return new CellComment(
+        uuid,
+        range,
+        author,
+        createdAt,
+        content,
+    )
+}
+
+// Comment Handler is in charge of comments for a particular cell
+export class CommentHandler extends UIMessageTarget {
+    private comments: Record<CommentID, CellComment> = {}; // holds all comments (root and replies)
+    private commentUIs: Record<string, CommentUI> = {};    // lookup table for comment UIs (keys are ranges, serialized as strings)
+    private commentRoots: Record<string, CommentID> = {};  // lookup table for range (serialized as string) to root comments
+    private highlights: Record<string, string[]> = {};     // deltaDecorations, range string -> monaco decoration ids.
+
+    constructor(readonly cellId: number, readonly editor: IStandaloneCodeEditor) {
+        super();
+
+        this.subscribe(CreateComment, (cellId, comment) => {
+            this._add(comment);
+        });
+        this.subscribe(UpdateComment, (cellId, commentId, range, content) => {
+            this._update(commentId, range, content);
+        });
+        this.subscribe(DeleteComment, (cellId, commentId) => {
+            this._delete(commentId);
+        });
+
+        this.editor.onDidChangeModelContent((evt: IModelContentChangedEvent) => {
```

Comment:
This code is responsible for updating comment ranges based on edits to the cell. I played around with a bunch of ways to do this. I landed on this way, which basically piggybacks on Monaco's range highlights which already do the "right thing". It seemed a bit ugly to rely on on this functionality, but on the other hand it turned out to be the easiest for me to reason about. 

100% open to suggestions on how to make it better if you have any ideas for APIs, etc., I might have missed.